File: src/DbFs.php
<?php

namespace Ething;


// gridfs helpers

class DbFs
{
    
	private $ething;
	
	public function __construct(Ething $ething){
		
		$this->ething = $ething;
		
	}
	
	public function storeFile($filename, $contents, $metadata = array()){
		if(!empty($contents)){
			$bucket = $this->ething->db()->selectGridFSBucket();
			
			$stream = $bucket->openUploadStream($filename, array(
				'metadata' => $metadata
			));
			
			fwrite($stream, $contents);
			$id = (string)$bucket->getFileIdForStream($stream);
			fclose($stream);
			
			return $id;
		}
		
		return null;
	}
	
	public function retrieveFile($id){
		if(isset($id)){
			$bucket = $this->ething->db()->selectGridFSBucket();
			try {
				$stream = $bucket->openDownloadStream(new \MongoDB\BSON\ObjectId($id));
				return stream_get_contents($stream);
			} catch(\MongoDB\GridFS\Exception\FileNotFoundException $e){}
		}
		
		return null;
	}
	
	public function removeFile($id){
		if(isset($id)){
			$bucket = $this->ething->db()->selectGridFSBucket();
			try {
				$bucket->delete(new \MongoDB\BSON\ObjectId($id));
			} catch(\MongoDB\GridFS\Exception\FileNotFoundException $e){}
		}
	}
	
	public function getFileMetadata($id){
		$metadata = array();
		
		if(isset($id)){
			$bucket = $this->ething->db()->selectGridFSBucket();
			try {
				$stream = $bucket->openDownloadStream(new \MongoDB\BSON\ObjectId($id));
				$metadata = $bucket->getFileDocumentForStream($stream);
			} catch(\MongoDB\GridFS\Exception\FileNotFoundException $e){}
		}
		
		return $metadata;
	}
	
	public function getFileSize($id){
		if(isset($id)){
			$bucket = $this->ething->db()->selectGridFSBucket();
			try {
				$stream = $bucket->openDownloadStream(new \MongoDB\BSON\ObjectId($id));
				$metadata = $bucket->getFileDocumentForStream($stream);
				return $metadata['length'];
			} catch(\MongoDB\GridFS\Exception\FileNotFoundException $e){}
		}
		
		return 0;
	}
	
}

File: ething/DbFs.py
# coding: utf-8


# gridfs helpers

import gridfs


class DbFs(object):

    def __init__(self, db):

        self.db = db

    def storeFile(self, filename, contents, metadata=None):
        if contents:

            bucket = gridfs.GridFSBucket(self.db)

            grid_in = bucket.open_upload_stream(filename, metadata=metadata)

            grid_in.write(contents)
            grid_in.close()

            return grid_in._id

        return None

    def retrieveFile(self, id):
        content = None
        if id:
            bucket = gridfs.GridFSBucket(self.db)
            try:
                grid_out = bucket.open_download_stream(id)
                content = grid_out.read()

            except gridfs.errors.NoFile:
                pass

        return content

    def removeFile(self, id):
        if id:
            bucket = gridfs.GridFSBucket(self.db)
            try:
                bucket.delete(id)
            except gridfs.errors.NoFile:
                pass

    def getFileMetadata(self, id):
        metadata = {}

        if id:
            fs = gridfs.GridFS(self.db)
            file = fs.find_one(id)
            if file is not None:
                metadata = file.metadata

        return metadata

    def getFileSize(self, id):
        if id:
            fs = gridfs.GridFS(self.db)
            file = fs.find_one(id)
            if file is not None:
                return file.length

        return 0

    def listFiles(self):
        fs = gridfs.GridFS(self.db)
        return fs.list()
